### PR TITLE
CD-631 Migrate to declarative pipeline with docker agent

### PR DESCRIPTION
**Pull request by gbastkowski**
Open points:
* What is a good email address for the maintainer label in the Dockerfile

What to check:
* Green build on build-master
* Green build on ide1
* PDF generation must be checked
* Ideally, someone executes `build.sh mvn clean install` on her/his local machine. But this will take a while

### CODE REVIEW

**Comment by yoosiba** [comment]
any specific settings need to be done before running the `build.sh` ?
I got this:

```shell
TMP_WS="ws_tmp"
mkdir $TMP_WS
OPTIONS="\
-Dmaven.test.failure.ignore
-e
-DWORKSPACE=$TMP_WS"

./build.sh mvn -U clean package -PbuildProduct ${OPTIONS}
(...)
[WARNING] Error initializing: org.eclipse.tycho.p2.resolver.P2DependencyResolver@176996c3
java.lang.RuntimeException: java.lang.IllegalStateException: Service is not registered class='interface org.eclipse.tycho.core.shared.ProxyServiceFacade'
    at org.eclipse.sisu.equinox.embedder.internal.DefaultEquinoxEmbedder.checkStarted (DefaultEquinoxEmbedder.java:312)
    at org.eclipse.sisu.equinox.embedder.internal.DefaultEquinoxEmbedder.getService (DefaultEquinoxEmbedder.java:286)
    at org.eclipse.sisu.equinox.embedder.internal.DefaultEquinoxEmbedder.getService (DefaultEquinoxEmbedder.java:280)
(...)
[ERROR] Internal error: java.lang.RuntimeException: Could not instantiate required component: com.google.inject.ProvisionException: Unable to provision, see the following errors:
[ERROR]
[ERROR] 1) Error notifying InjectionListener org.eclipse.sisu.plexus.PlexusBeanBinder@411c6d44 of org.eclipse.tycho.p2.resolver.P2DependencyResolver.
[ERROR]  Reason: java.lang.RuntimeException: java.lang.IllegalStateException: Service is not registered class='interface org.eclipse.tycho.core.shared.ProxyServiceFacade'
[ERROR]   while locating org.eclipse.tycho.p2.resolver.P2DependencyResolver
[ERROR]   at ClassRealm[extension>org.eclipse.tycho:tycho-maven-plugin:1.0.0, parent: sun.misc.Launcher$AppClassLoader@74a14482] (via modules: org.eclipse.sisu.wire.WireModule -> org.eclipse.sisu.plexus.PlexusBindingModule)
[ERROR]   while locating org.eclipse.tycho.core.DependencyResolver annotated with @com.google.inject.name.Named(value=p2)
[ERROR]
[ERROR] 1 error
[ERROR]       role: org.eclipse.tycho.core.DependencyResolver
[ERROR]   roleHint: p2
[ERROR] -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: Could not instantiate required component
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:122)
(...)
[ERROR]
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/InternalErrorException
```

**Comment by yoosiba** [comment]
maybe it is related to the strange location of created m2 repo: `./?/m2/repository/...`

**Comment by mor-n4** [comment]
please ensure our current "extended" builds are still working with these changes

**Comment by mor-n4** [comment]
have you updated the design document? (esp. Chapter 4)

File: build.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+# build docker image
+docker build --build-arg MAVEN_CENTRAL_URL=$MAVEN_CENTRAL_URL -t n4js-build docker-build/
+
+# run the docker container
+docker run -ti --rm      \
+    -u `id -u`:`id -g`   \
+    -v ~/.m2:/root/.m2/  \
+    -v $(pwd):/workspace \
+    n4js-build $@
```
Comment (mor-n4) [inline build.sh]: very clean, readable formatting in all these scripts. Nice!

File: docker-build/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ENTRYPOINT ["/usr/local/bin/mvn-entrypoint.sh"]
+CMD ["mvn"]
+
+#
+# Maven installation:
+# see https://github.com/carlossg/docker-maven/tree/ecf54b9839caed8aa2bcf9b8f7bb19594634ee89/jdk-8
+# and https://hub.docker.com/_/maven/
+#
+ARG MAVEN_VERSION=3.5.2
+ARG USER_HOME_DIR="/root"
```
Comment (yoosiba) [inline docker-build/Dockerfile]: can this be a reason my m2 repo is in `./?/m2/repository/`

File: docker-build/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,111 @@
+FROM openjdk:8u151-jdk
+
+LABEL version="0.1.0" \
+      description="Build n4js with Maven. Includes xvfb, node." \
+      maintainer="Continuous Delivery Team <staff-build-release@numberfour.eu>" \
```
Comment (yoosiba) [inline docker-build/Dockerfile]: maybe `n4js-dev@eclipse.org` ?

File: docker-build/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      eu.numberfour.dockerimage.name="numberfour/build-n4js" \
```
Comment (yoosiba) [inline docker-build/Dockerfile]: definitely not `eu.numberfour`
how are those labels used?

**Comment by gbastkowski** [comment]
Please review again.
Local build with Docker should work now.

**Comment by mor-n4** [comment]
@torstenkraemer Since you are continuing with this on a different branch now, can this PR be closed? (to avoid confusion)

**Comment by mor-n4** [comment]
Superseded by PR #387.

**Pull request closed without merging**